### ISSUE TITLE
How to scale virtual kubernetes, can we put load balancer for virtual nodes?

### ISSUE THREAD

**Issue by AfreenShaha** [virtual-kubelet/virtual-kubelet#409]
---

### Environment summary

Provider (e.g. ACI, AWS Fargate, Hyper)
ACI

Version (e.g. 0.1, 0.2-beta)

K8s Master Info (e.g. AKS, ACS, Bare Metal, EKS)
Azure Kubernetes Service
Install Method (e.g. Helm Chart, )
Helm
### Issue Details
Can we put a load balancer in yaml for virtual kubernetes pods?
Can we scale the virtual nodes on kubernetes?
Do we need to do scale  virtual kubernete nodes using below help document mentioned in the msdn?
https://docs.microsoft.com/en-us/azure/aks/autoscaler
how to create multiple virtual  windows nodes in a virtual kubernetes?

**Comment by erjadi** [comment]
Hi AfreenShaha,

Yes you can load balance over ACI-backed pods but there are a few caveats:
- When deploying a LoadBalancer on AKS, it will actually deploy an Azure Load Balancer.
- Unfortunately a LoadBalancer cannot balance over non-VM resources, however the way networking works by default in AKS is that you can still reach your ACI pods through the overlay network.


Long story short:
- Loadbalancing works as normal, BUT you have to deploy the loadbalancer before you install VK, otherwise the loadbalancer deployment will fail.

Anyone please correct me if I'm wrong, but I believe this issue is still open (see https://github.com/virtual-kubelet/virtual-kubelet/issues/41)

**Comment by robbiezhang** [comment]
Thanks Eric for the answer. 
#41 is mitigated in AKS 1.9+ clusters. Loadbalancer service should work well with the VK now.
@AfreenShaha, as Eric mentioned, you need a load balancer service in the AKS cluster in front of the VK pods. You can follow the [k8s doc ](https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/) to create the service and expose the pods.

For the virtual node scaling, it's different from scaling AKS nodes. You don't need to scale if you are out of ACI capacity. Instead, you need to file an Azure support ticket to get more ACI quota.
Here are the scenarios that you might need to add virtual nodes when you want the virtual node managed ACI container group running in
- different OS type (Linux, Windows)
- different Azure subscription/resource group
- different Azure region

Providing the above facts, to install a new virtual node in your cluster, you just need to give a different set of parameters when running the helm chart. [Reference](https://github.com/virtual-kubelet/virtual-kubelet/blob/master/providers/azure/README.md#manual-set-up) However, it is preferred to customize the following parameters if you want to run:
- [Required]: different release name
- [Required]: different node name
- [Optional]: different Azure subscription or resource group for better management.

**Comment by AfreenShaha** [comment]
Thank you for a quick reply. 
As per the discussion here is the flow, please correct me If I am wrong:
-Create azure kubernetes service
-Create the Load balancer
-Create azure container instance with windows as OS(as I have ASP.Net app deployed on windows container)
-Deploy the application YAML(I took the yaml definition reference  from here : https://docs.microsoft.com/en-us/azure/aks/virtual-kubelet)

**Few questions:**
-Can you please tell me how to check the ACI capacity?
-We are working on simulator, and each pod needs to simulate 500 devices using 500 threads in it.
  like this we need to start 100 pods to start 50000 threads at once.
-When I go to azure portal, I see the CPU usage of virtual kubernete node as 0%(as shown in image). I need help to see how many pods ACI can support?
-Can we use this command to autoscale the number of pods in virtual kubernetes node?
"kubectl autoscale deployment appname --cpu-percent=50 --min=3 --max=10"
-"kubectl get hpa" returns below error for virtual kubernetes, can you please help what can be wrong here?
"the server doesn't have a resource type "hpa""

<img width="825" alt="azure_kubernetes_issues" src="https://user-images.githubusercontent.com/15724133/48399191-fa6e6300-e748-11e8-86b1-1f847ad8a665.png">

**Comment by robbiezhang** [comment]
@AfreenShaha ACI for Windows doesn't yet report metrics, i.e. CPU, memory, etc. nor multi-container. So, your scenario will not work on the virtual node with ACI. There is a plan to support them in Windows Server 2019 containers. However, no ETA currently.

You can use the ARMClient tool to get the ACI capacity of your subscription in a particular region.
```bash
armclient get /subscriptions/{subscriptionId}/providers/Microsoft.ContainerInstance/locations/{region}/usages?api-version=2018-10-01
```
You can get the ARMClient tool from the following 2 repositories:
https://github.com/projectkudu/ARMClient
https://github.com/yangl900/armclient-go

**Comment by AfreenShaha** [comment]
Thanks @robbiezhang  for a quick response on this.
Just to confirm, when you said scenario will not work, I believe you are talking about running 100 pods in a ACI? please correct me if I am wrong?

**Comment by robbiezhang** [comment]
Nope, you can run 100 pods in ACI in a subscription without any problem. If you want to run more, you can request to increase the quota for your subscription.
The scenario I mentioned here is the HPA, since the ACI Windows container group doesn't report CPU/memory metrics.

**Comment by AfreenShaha** [comment]
Thanks robbiezhang. If azure kubernetes directly support the deployment of windows containers then  it would be great.
Currently it just supports the linux containers.
